crates/agglayer-prover/src/rpc.rs

```diff
@@ -14,11 +33,85 @@ impl ProofGenerationService for ProverRPC {
+        let mut executor = self.executor.clone();
+        let executor = executor
+            .ready()
+            .await
+            .map_err(|_error| tonic::Status::internal("Unable to get proof executor"))?;
+
+        match executor
+            .call(Request {
+                initial_state,
+                batch_header,
+            })
+            .await
+        {
+            Ok(result) => {
+                let response = agglayer_prover_types::v1::ProofGenerationResponse {
+                    proof: bincode::options().serialize(&result.proof).map_err(|_| {
+                        tonic::Status::internal("Unable to serialize generated proof")
+                    })?,
+                };
+
+                PROVING_REQUEST_SUCCEEDED.add(1, metrics_attrs);
+                return Ok(tonic::Response::new(response));
+            }
+            Err(error) => {
+                if let Some(error) = error.downcast_ref::<Error>() {
+                    error!("Failed to generate proof: {}", error);
+                    let response = match error {
+                        Error::UnableToExecuteProver => {
+                            tonic::Status::internal("Unable to execute prover")
+                        }
+                        Error::ExecutorFailed(error) => {
+                            if let Ok(bytes_error) = default_bincode_options().serialize(&error) {
+                                tonic::Status::with_details(
+                                    tonic::Code::InvalidArgument,
+                                    error.to_string(),
+                                    bytes_error.into(),
+                                )
+                            } else {
+                                warn!("Unable to serialize Execution  error: {}", error);
+                                tonic::Status::invalid_argument(error.to_string())
+                            }
+                        }
+                        Error::ProofVerificationFailed(error) => {
+                            if let Ok(bytes_error) = default_bincode_options().serialize(&error) {
+                                tonic::Status::with_details(
+                                    tonic::Code::InvalidArgument,
+                                    error.to_string(),
+                                    bytes_error.into(),
+                                )
+                            } else {
+                                warn!("Unable to serialize SP1 verification error: {}", error);
+                                tonic::Status::invalid_argument(error.to_string())
+                            }
+                        }
+                        Error::ProverFailed(_) => tonic::Status::internal(error.to_string()),
+                    };
+
+                    return Err(response);
+                } else {
+                    error!("Failed to generate proof: {:?}", error);
+
+                    PROVING_REQUEST_FAILED.add(1, metrics_attrs);
```

Comment:
If I got the if-else nesting correct, this counter is only increased if the error is not a `agglayer_prover_types::Error`. Is that intentional? Would another metric that covers `agglayer_prover_types::Error` be useful?

Reply:
Ah yes indeed, I will move this one above the if.